### ISSUE TITLE
cc-options do not get passed to GHC

### ISSUE THREAD

**Issue by bitonic** [haskell/cabal#4435]
`cc-options` currently do not get seem to get passed to GHC. It would be useful if they were, since GHC is capable of compiling C/C++ files and now it might more than before given the addition of `addForeignFile` in GHC 8.2, which lets users compile and link C/C++ files from TH.

**Comment by bitonic** [comment]
Note that in practice one needs to put the cc options both in `cc-options` _and_ in `ghc-options` via `-optc`, since inline-c is often used with hsc2hs which does get the flags from `cc-options`.

**Comment by bitonic** [comment]
Note that the same happens for linking: `extra-libraries` are not passed to GHC, which makes `loadDLL` fail when it `dlopen`s shared libraries referring to any of the specified extra-libraries.

**Comment by bitonic** [comment]
Note that `cabal` even emits warnings like

```
Warning: Instead of 'ghc-options: -lceres ...' use 'extra-libraries:
ceres ...'
```

even if they do not have the same effect.

**Comment by bitonic** [comment]
Another note: also in the case of `extra-libraries` one needs to have both `ghc-options: -lfoo` and `extra-libraries: foo`, since GHC might load with `dlopen` but cabal is still the one doing the final linkage

**Comment by ezyang** [comment]
Related issue #3700

Not passing extra-libraries to GHC as `-l` is probably a separate bug. If this doesn't cause any regressions, we probably should do this, at the very least when `other-extensions: TemplateHaskell` is specified, and maybe unconditionally.

**Comment by ezyang** [comment]
@bitonic If you're interested in solving this ticket I'd be happy to advise.

**Comment by basvandijk** [comment]
I just run into this today when building [haskell-opencv](https://github.com/LumiGuide/haskell-opencv) with cabal-install. I got the following error:

```
<no location info>: error:
    ghc: panic! (the 'impossible' happened)
  (GHC version 8.4.4 for x86_64-unknown-linux):
	Loading temp shared object failed: 
          /run/user/1001/ghc14715_0/libghc_95.so: 
            undefined symbol: _Znwm
```

Note that `_Znwm` demangles to: `operator new(unsigned long)`.

After adding `ghc-options: -lstdc++` to `opencv.cabal` the build succeeded. 

Note that I did get the warning: 
`Warning: Instead of 'ghc-options: -lstdc++' use 'extra-libraries: stdc++'`

**Comment by phadej** [comment]
For the record, `haskell-opencv` does specify: 

```
    extra-libraries:   stdc++
```

---

I agree with Edward, patch is probably easy to write (and if CI is green, to merge).

**Comment by domenkozar** [comment]
It's the same issue with `cxx-options: -std=c++14`, it still needs `ghc-options: -optc-std=c++14`

**Comment by phadej** [comment]
Is this similar problem as with `ld-options` as with @nh2 recently fixed?

**Comment by ulysses4ever** [comment]
This issue was marked high-priority a while ago. Since then, there were a number of updates around passing options between cabal and GHC. To determine whether these updates had any impact in the issue, a minimal reproducer would be helpful. 

Do we have a minimal reproducer?

**Comment by phadej** [comment]
https://github.com/haskell/cabal/issues/9801

**Comment by cblp** [comment]
@ulysses4ever here's a minimal repro https://github.com/cblp/cabal-cxx-options

**Comment by ulysses4ever** [comment]
@cblp thanks! any chance you could check that cabal from https://github.com/haskell/cabal/pull/10969 fixes the problem? (you'd have to build it locally). If it does, keep in mind that #10969 may require help from interested 3rd parties (e.g. you!) to get into a mergeable state.

**Comment by cblp** [comment]
@ulysses4ever yes, the proposed version passes my tests, even after I added some additional checks
